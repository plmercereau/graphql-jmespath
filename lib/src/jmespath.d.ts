declare module 'jmespath' {
    export type ObjectType = any // TODO
    export type JmespathComparator = 'EQ' | 'NE' | 'GT' | 'LT' | 'LTE' | 'GTE'
    //   TODO check type - which properties are optional?
    export type ASTNode = {
        value: any // TODO
        type: string
        name: string
        children: Array<ASTNode>
        jmespathType?: any // TODO
    }

    export function compile(expression: string): ASTNode
    export function search(obj: ObjectType, expression: string): ObjectType
}
