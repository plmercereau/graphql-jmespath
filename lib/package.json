{
    "name": "jmespath-graphql",
    "author": "Pierre-Louis Mercereau",
    "version": "0.1.0",
    "license": "MIT",
    "main": "dist/index.cjs.js",
    "module": "dist/index.esm.js",
    "types": "dist/index.d.ts",
    "source": "src/index.ts",
    "files": [
        "dist",
        "umd",
        "README.md"
    ],
    "exports": {
        ".": {
            "import": {
                "node": "./dist/index.cjs.js",
                "default": "./dist/index.esm.js"
            },
            "require": "./dist/index.cjs.js"
        }
    },
    "publishConfig": {
        "access": "public"
    },
    "scripts": {
        "dev": "vite build --watch",
        "build": "run-p build:lib build:umd",
        "build:lib": "vite build",
        "build:umd": "vite build --config vite.umd.config.js",
        "test": "vitest run",
        "test:watch": "vitest"
    },
    "devDependencies": {
        "c8": "^7.11.3",
        "jsdom": "^20.0.0",
        "typescript": "^4.5.4",
        "vite": "^2.8.0",
        "vitest": "^0.17.0"
    },
    "dependencies": {
        "graphql": "^16.5.0",
        "deepmerge": "^4.2.2",
        "dot-prop": "^7.2.0",
        "graphql-request": "^4.2.0",
        "jmespath": "^0.16.0",
        "json-to-graphql-query": "^2.2.3"
    }
}
