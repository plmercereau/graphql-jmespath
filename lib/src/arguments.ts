import { JmespathComparator } from 'jmespath'
export type WhereArgumentPath = (initialPath: string, subPath: string, comparator: JmespathComparator) => string