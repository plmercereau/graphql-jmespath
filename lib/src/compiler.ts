import { ASTNode } from 'jmespath'
import merge from 'deepmerge'
import { setProperty, getProperty } from 'dot-prop'

export const recursiveJmespathToObject = (
    node: ASTNode,
    path: string = '',
    // TODO do we really need to keep track of wildcards in the compiler, now that it is tracked in the post ast->object compilation phase?
    wildcard = true
) => OPERATIONS[node.type](node, path, wildcard)

const joinPaths = (left?: string, right?: string) =>
    left && right ? `${left}.${right}` : left || right || ''

type OperationResult = {
    value: any
    path: string
    wildcard: boolean
}
type OperationFunction = (
    node: ASTNode,
    path: string,
    wildcard: boolean
) => OperationResult

function mergeProperty<ObjectType extends Record<string, any>>(
    object: ObjectType,
    path: string,
    value: any
): ObjectType {
    const oldValue: any = getProperty(object, path)
    const newValue = merge(oldValue, value)
    return setProperty(object, path, newValue)
}

const Merge: OperationFunction = (node, path, wc) => {
    const [left, right] = node.children
    const leftResult = recursiveJmespathToObject(left, path, wc)
    const rightResult = recursiveJmespathToObject(right, path, wc)
    const wildcard = wc || leftResult.wildcard || rightResult.wildcard
    return {
        path: joinPaths(leftResult.path, rightResult.path),
        value: merge(leftResult.value, rightResult.value),
        wildcard
    }
}

const Identity: OperationFunction = (_, path, wildcard) => {
    return {
        path,
        value: true,
        wildcard
    }
}

const FirstChild: OperationFunction = (node, path, wildcard) =>
    recursiveJmespathToObject(node.children[0], path, wildcard)

const MultiSelect: OperationFunction = (node, path, wc) => {
    let value = {}
    let wildcard = wc
    for (const child of node.children) {
        const res = recursiveJmespathToObject(child, path, wc)
        value = merge(value, res.value)
        wildcard = wildcard || res.wildcard
    }
    return { value, path, wildcard }
}

const OPERATIONS: Record<string, OperationFunction> = {
    // * Operations defined in JMESPath's abstract syntax, but not yet encountered in the tests: Index, Slice
    Field: (node, _, wildcard) => {
        return { value: { [node.name]: true }, path: node.name, wildcard }
    },
    Subexpression: (node, path, wc) => {
        const [left, right] = node.children
        const leftResult = recursiveJmespathToObject(left, path, wc)
        const rightResult = recursiveJmespathToObject(right, path, wc)
        const wildcard = wc || leftResult.wildcard || rightResult.wildcard
        if (leftResult.path) {
            const result = {
                value: {},
                path: joinPaths(leftResult.path, rightResult.path),
                wildcard
            }
            setProperty(result.value, leftResult.path, rightResult.value)
            return result
        } else {
            return rightResult
        }
    },
    OrExpression: (node, path, wc) => {
        const [left, right] = node.children
        const leftResult = recursiveJmespathToObject(left, path, wc)
        const rightResult = recursiveJmespathToObject(right, path, wc)
        const wildcard = wc || leftResult.wildcard || rightResult.wildcard
        return {
            path,
            value: merge(leftResult.value, rightResult.value),
            wildcard
        }
    },
    Comparator: Merge,
    AndExpression: Merge,
    FilterProjection: (node, path, wc) => {
        const [first, second, third] = node.children
        const firstResult = recursiveJmespathToObject(first, path, wc)
        const secondResult = recursiveJmespathToObject(second, path, wc)
        const thirdResult = recursiveJmespathToObject(third, path, wc)
        const wildcard =
            wc ||
            firstResult.wildcard ||
            secondResult.wildcard ||
            thirdResult.wildcard
        const value = firstResult.value
        if (firstResult.path) {
            setProperty(
                value,
                firstResult.path,
                merge(secondResult.value, thirdResult.value)
            )
        }
        return {
            path: firstResult.path,
            value,
            wildcard
        }
    },
    Identity,
    Literal: Identity,
    Projection: (node, path, wc) => {
        const [left, right] = node.children
        const leftResult = recursiveJmespathToObject(left, path, wc)
        const rightResult = recursiveJmespathToObject(right, path, wc)
        const wildcard = wc || leftResult.wildcard || rightResult.wildcard
        const result = {
            value: leftResult.value,
            path: joinPaths(leftResult.path, rightResult.path),
            wildcard
        }
        // ! right.type !== 'Identity' is hacky, but it works so far in all the tests
        if (leftResult.path && right.type !== 'Identity') {
            mergeProperty(result.value, leftResult.path, rightResult.value)
        }
        return result
    },
    Flatten: FirstChild,
    IndexExpression: FirstChild,
    Pipe: (node, path, wc) => {
        const [left, right] = node.children
        const leftResult = recursiveJmespathToObject(left, path, wc)
        const rightResult = recursiveJmespathToObject(right, path, wc)
        const wildcard = wc || leftResult.wildcard || rightResult.wildcard
        if (rightResult.path && leftResult.path) {
            const newPath = `${leftResult.path}.${rightResult.path}`
            const result = {
                value: leftResult.value,
                path: newPath,
                wildcard
            }
            if (!getProperty(leftResult.value, newPath)) {
                // * Only expressions like 'foo | bar' but not like 'people[?general.id==`100`] | [0].general'
                setProperty(result.value, leftResult.path, rightResult.value)
            }

            return result
        } else if (left.type === 'MultiSelectHash') {
            // * '{"x": foo, "y": bar} | [y.baz]'
            const value = {}
            for (const leftChild of left.children) {
                const rightValue =
                    rightResult.value['*'] || rightResult.value[leftChild.name]
                if (rightValue) {
                    const realPath = recursiveJmespathToObject(
                        leftChild,
                        path,
                        wc
                    ).path
                    setProperty(value, realPath, rightValue)
                }
            }
            return { value, path: '', wildcard }
        } else if (left.type === 'OrExpression') {
            // * 'foo.bam || bar | baz'
            const value = {}
            for (const leftChild of left.children) {
                const realPath = recursiveJmespathToObject(
                    leftChild,
                    path,
                    wc
                ).path
                setProperty(value, realPath, rightResult.value)
            }
            return {
                value,
                path,
                wildcard: wildcard
            }
        } else {
            // * 'foo | other || bar'
            if (right.type === 'OrExpression') {
                setProperty(
                    leftResult.value,
                    leftResult.path,
                    rightResult.value
                )
            }
            return {
                value: leftResult.value,
                path,
                wildcard: wildcard
            }
        }
    },
    MultiSelectList: MultiSelect,
    MultiSelectHash: MultiSelect,
    KeyValuePair: (node, path, wildcard) => {
        return recursiveJmespathToObject(node.value, path, wildcard)
    },
    ValueProjection: (node, path, wc) => {
        const [left, right] = node.children
        const leftResult = recursiveJmespathToObject(left, path, wc)
        const rightResult = recursiveJmespathToObject(right, path, wc)
        const newPath = joinPaths(leftResult.path, '*')
        const value = {}
        setProperty(value, newPath, rightResult.value)
        return {
            value,
            path: joinPaths(newPath, rightResult.path),
            wildcard: true
        }
    },
    Function: (node, path, wc) => {
        // ? check wildcard in children
        let root = {}
        let rootPath = ''
        let children = {}
        for (const child of node.children) {
            const result = recursiveJmespathToObject(child, path, wc)
            if (child.type === 'ExpressionReference') {
                children = merge(children, result.value)
            } else if (child.type !== 'Literal') {
                // * Skip literals
                root = merge(root, result.value)
                rootPath = result.path || ''
            }
        }
        if (rootPath && Object.keys(children).length) {
            setProperty(root, rootPath, children)
        }
        return { value: root, path: rootPath, wildcard: wc }
    },
    ExpressionReference: FirstChild,
    NotExpression: FirstChild,
    Current: (_, path, wildcard) => {
        return { value: {}, path, wildcard }
    }
}
