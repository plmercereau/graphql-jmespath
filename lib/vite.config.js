import fs from 'fs'
import path from 'path'

import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'
import tsconfigPaths from 'vite-tsconfig-paths'

const PWD = process.env.PWD
const pkg = require(path.join(PWD, 'package.json'))

const tsEntry = path.resolve(PWD, 'src/index.ts')
const entry = fs.existsSync(tsEntry) ? tsEntry : tsEntry.replace('.ts', '.tsx')

const deps = [
    ...Object.keys(Object.assign({}, pkg.peerDependencies, pkg.dependencies))
]

export default defineConfig({
    plugins: [
        tsconfigPaths(),
        dts({
            exclude: ['**/*.spec.ts', '**/*.test.ts', '**/tests/**'],
            afterBuild: () => {
                const types = fs.readdirSync(path.join(PWD, 'dist/src'))
                types.forEach((file) => {
                    fs.renameSync(
                        path.join(PWD, 'dist/src', file),
                        path.join(PWD, 'dist', file)
                    )
                })
                fs.rmdirSync(path.join(PWD, 'dist/src'))
            }
        })
    ],
    test: {
        globals: true,
        environment: 'jsdom',
        reporters: 'verbose',
        include: [
            `${PWD}/src/**/*.{spec,test}.{ts,tsx}`,
            `${PWD}/tests/**/*.{spec,test}.{ts,tsx}`
        ],
        // Note: temporarily disabled threads, because of a bug in vitest
        // https://github.com/vitest-dev/vitest/issues/1171
        threads: false,
        coverage: {
            enabled: process.env.CI === 'true',
            reporter: ['json']
        }
    },
    build: {
        sourcemap: true,
        lib: {
            entry,
            name: pkg.name,
            fileName: (format) =>
                format === 'cjs' ? `index.cjs.js` : `index.esm.js`,
            formats: ['cjs', 'es']
        },
        rollupOptions: {
            external: (id) => deps.some((dep) => id.startsWith(dep)),
            output: {
                globals: {
                    graphql: 'graphql',
                    react: 'React',
                    'react-dom': 'ReactDOM',
                    'react/jsx-runtime': '_jsx'
                }
            }
        }
    }
})
