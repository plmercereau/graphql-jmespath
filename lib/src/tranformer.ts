import { compile, search } from 'jmespath'
import { jsonToGraphQLQuery } from 'json-to-graphql-query'
import { recursiveJmespathToObject } from './compiler'
import { assertName, GraphQLFieldMap, GraphQLSchema } from 'graphql/type'
import { DocumentNode, parse } from 'graphql'

type JsonObject = { [key: string]: JsonObject | boolean }

type UnknownFieldsOption = 'ignore' | 'raise'
type StopAtPathOption = ((path: string) => boolean) | string | string[]

type ArgumentValue = boolean | string | number | null | undefined

type GraphQLArguments = {
    [key: string]: GraphQLArguments | ArgumentValue | ArgumentValue[]
}

type Options = {
    schema?: GraphQLSchema
    stopAtPath?: StopAtPathOption
    rootQuery?: string
    rootArgs?: GraphQLArguments
    unknownFields?: UnknownFieldsOption
}

export { search }

export const expressionToObject = (
    expression: string
): JsonObject | boolean => {
    const ast = compile(expression)
    return recursiveJmespathToObject(ast).value
}

export const expressionToJSONQuery = (
    expression: string,
    {
        schema,
        stopAtPath = [],
        unknownFields = 'ignore',
        rootQuery,
        rootArgs
    }: Options = {}
) => {
    const json = filterObjectPaths(
        expressionToObject(expression),
        { stopAtPath, unknownFields, schema },
        {
            // TODO differs when rootQuery is provided
            node: schema?.getQueryType()
        }
    )
    if (typeof json === 'boolean' || !Object.keys(json).length) {
        throw Error('Empty query')
    }
    const query = rootQuery
        ? {
              [rootQuery]: { __args: rootArgs, ...json }
          }
        : json
    return { query }
}

export const expressionToGraphQL = (
    expression: string,
    { pretty = false, ...options }: { pretty?: boolean } & Options = {}
): string => {
    const json = expressionToJSONQuery(expression, options)
    return jsonToGraphQLQuery(json, { pretty })
}
export const expressionToGraphQLDocument = (
    expression: string,
    options?: Options
): DocumentNode => {
    // TODO option: validate against schema
    return parse(expressionToGraphQL(expression, options))
}
// export class Expression {
//     readonly expression: string
//     rootQuery?: string
//     rootArgs?: GraphQLArguments
//     stopAtPath: StopAtPathOption
//     unknownFields: UnknownFieldsOption
//     schema?: GraphQLSchema
//     readonly ast: ASTNode

//     constructor(
//         expression: string,
//         {
//             schema,
//             stopAtPath = [],
//             rootQuery,
//             rootArgs,
//             unknownFields = 'ignore'
//         }: Options = {}
//     ) {
//         this.expression = expression
//         this.rootQuery = rootQuery
//         this.rootArgs = rootArgs
//         this.stopAtPath = stopAtPath
//         this.unknownFields = unknownFields
//         this.schema = schema
//         this.ast = compile(expression)
//     }

//     toObject(): JsonObject | boolean {
//         return recursiveJmespathToObject(this.ast).value
//     }

//     toJSONQuery(options: Options = {}) {
//         const { schema, stopAtPath, unknownFields, rootQuery, rootArgs } = {
//             ...this,
//             ...options
//         }
//         const json = filterObjectPaths(
//             this.toObject(),
//             { stopAtPath, unknownFields, schema },
//             {
//                 // TODO differs when rootQuery is provided
//                 node: schema?.getQueryType()
//             }
//         )
//         if (typeof json === 'boolean' || !Object.keys(json).length) {
//             throw Error('Empty query')
//         }
//         const query = rootQuery
//             ? {
//                   [rootQuery]: { __args: rootArgs, ...json }
//               }
//             : json
//         return { query }
//     }

//     toGraphQLDocument(options: Options = {}) {
//         // TODO option: validate against schema
//         return parse(this.toGraphQL(options))
//     }

//     toGraphQL({
//         pretty = false,
//         ...options
//     }: { pretty?: boolean } & Options = {}): string {
//         const json = this.toJSONQuery(options)
//         return jsonToGraphQLQuery(json, { pretty })
//     }

//     search(obj: any) {
//         return search(obj, this.expression)
//     }
// }

// ? move to class ?
// TODO (big): compile jmespath syntax such as filter, sort, etc into Hasura "_where", "_orderby", "_limit", "_offset"...
const filterObjectPaths = (
    obj: JsonObject | boolean,
    options: Required<Pick<Options, 'stopAtPath' | 'unknownFields'>> &
        Pick<Options, 'schema'>,
    {
        node,
        path = ''
    }: {
        node?: any // TODO type
        path?: string
    } = {}
): JsonObject | boolean => {
    const { stopAtPath, unknownFields, schema } = options
    if (schema) {
        // * When we have a schema, we only add parts of the graph that actually exist in the schema
        if (!node) {
            return false
        }

        // ! Ignore sub-obkect if one of its args is required
        if (node.args?.some((arg: any) => arg.type.toString().endsWith('!'))) {
            // * Note: in GraphQL, any type ending with ! is required
            return false
        }
    }
    // * We reached a leaf node, so we return the boolean value
    if (typeof obj === 'boolean') {
        return obj
    }

    // * stopAtPath: stop browsing further down the tree
    if (
        (typeof stopAtPath === 'function' && stopAtPath(path)) ||
        (typeof stopAtPath === 'string' && path === stopAtPath) ||
        (Array.isArray(stopAtPath) && stopAtPath.includes(path))
    ) {
        return true
    }

    // * Get the base type of the node, for instance required(list(something)) => something
    // ? Would it work with unions or interfaces?
    let type = node?.type
    while (type && 'ofType' in type) {
        type = type.ofType
    }
    const childFields: GraphQLFieldMap<any, any> | undefined =
        node?._fields || type?._fields || undefined

    const objectResult = Object.entries(obj).reduce<JsonObject>(
        (aggr, [key, value]) => {
            let childType = childFields?.[key]?.type
            while (childType && 'ofType' in childType) {
                childType = childType.ofType
            }

            if (key === '*') {
                if (childFields) {
                    // TODO what happens when there is nothing to expand?
                    // * Expand wildcard fields from the GraphQL schema, when available
                    Object.entries(childFields).forEach(
                        ([childKey, childValue]) => {
                            const childObject = filterObjectPaths(
                                value,
                                options,
                                {
                                    path: path
                                        ? `${path}.${childKey}`
                                        : childKey,
                                    node: childValue
                                }
                            )
                            if (
                                childObject === true ||
                                Object.keys(childObject).length
                            ) {
                                // * Only add when the child object is not empty
                                aggr[childKey] = childObject
                            }
                        }
                    )
                } else {
                    // * The wildcard cannot be expanded: raise an error, or ignore it
                    if (unknownFields === 'raise') {
                        throw new Error(
                            `The expression contains wildcards than cannot be expanded at path: ${path}`
                        )
                    }
                }
                return aggr
            } else {
                // * Raise an error if the key is not a valid GraphQL name
                assertName(key)

                if (childFields && !node.name) {
                    // * The field is not defined in the GraphQL schema: raise an error, or ignore it
                    if (unknownFields === 'raise') {
                        throw new Error(
                            `The expression contains an unknown GraphQL field: ${key} at path: ${path}`
                        )
                    }
                } else {
                    const subObject = filterObjectPaths(value, options, {
                        node: childFields?.[key],
                        path: path ? `${path}.${key}` : key
                    })
                    if (subObject === true || Object.keys(subObject).length) {
                        // * Only add when the subObject is not empty
                        aggr[key] = subObject
                    }
                }
            }
            return aggr
        },
        {}
    )
    if (Object.keys(objectResult).length == 0) {
        // * Only return the object when it is not empty
        return false
    }
    return objectResult
}
