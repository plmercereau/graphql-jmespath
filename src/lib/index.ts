import { ASTNode, compile } from 'jmespath'
import { jsonToGraphQLQuery } from 'json-to-graphql-query'
import { recursiveJmespathToObject } from './compiler'
import { assertName } from 'graphql/type'

type JsonObject = { [key: string]: JsonObject | boolean }

type StopAtPathOption = ((path: string) => boolean) | string | string[]

type Options = {
    pretty?: boolean
    stopAtPath?: StopAtPathOption
}

// TODO (big): compile jmespath syntax such as filter, sort, etc into Hasura "_where", "_orderby", "_limit", "_offset"...
const filterObjectPaths = (
    obj: JsonObject | boolean,
    stopAtPath: StopAtPathOption,
    path = ''
): JsonObject | boolean => {
    // * We reached a leaf node, so we return the boolean value
    if (typeof obj === 'boolean') {
        return obj
    }

    // * stopAtPath: stop browsing further down the tree
    if (
        (typeof stopAtPath === 'function' && stopAtPath(path)) ||
        (typeof stopAtPath === 'string' && path === stopAtPath) ||
        (Array.isArray(stopAtPath) && stopAtPath.includes(path))
    ) {
        // TODO: at a later point, we could also detect a stop if we hit a litteral value in the GraphQL schema e.g. a JSON field
        return true
    }

    return (
        Object.entries(obj)
            // * Ignore wildcard fields
            // TODO: raise an error if there is a wildcard
            // TODO: at a later point, we should enable the option to expand wildcard fields from the GraphQL schema, when available
            .filter(([key]) => key !== '*')
            .reduce<JsonObject>((aggr, [key, value]) => {
                // * Raise an error if the key is not a valid GraphQL name
                assertName(key)
                aggr[key] = filterObjectPaths(
                    value,
                    stopAtPath,
                    path ? `${path}.${key}` : key
                )
                return aggr
            }, {})
    )
}

export const objectToGraphQL = (
    obj: JsonObject,
    // TODO add information about the root query (a starting point for the path). For instance, query { user(id) { profile { name } } }
    { pretty = true, stopAtPath = [] }: Options = {}
): string => {
    const query = filterObjectPaths(obj, stopAtPath)
    if (Object.keys(query).length === 0) {
        throw Error('Empty query')
    }
    return jsonToGraphQLQuery({ query }, { pretty })
}

export const jmespathToObject = (expression: string) => {
    const ast = compile(expression)
    return astToObject(ast)
}

export const jmespathToGraphQL = (
    expression: string,
    options: Options = {}
) => {
    const obj = jmespathToObject(expression)
    return objectToGraphQL(obj, options)
}

export const astToObject = (node: ASTNode) =>
    recursiveJmespathToObject(node).value
