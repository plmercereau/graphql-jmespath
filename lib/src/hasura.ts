import { JmespathComparator } from 'jmespath'
import { WhereArgumentPath } from './arguments'
import { Options } from './tranformer'

const HASURA_COMPARATORS: Record<JmespathComparator, string> = {
    EQ: '_eq',
    NE: '_neq',
    GT: '_gt',
    GTE: '_gte',
    LT: '_lt',
    LTE: '_lte'
}

// TODO multiple entries with Vite (and package.json?). Make import work with:
/*
 import { whereArgumentPath } from 'graphql-jmespath/hasura'
 import { options } from 'graphql-jmespath/hasura'
 */
export const whereArgumentPath: WhereArgumentPath = (
    initialPath,
    subPath,
    comparator
) => `${initialPath}.__args.where.${subPath}.${HASURA_COMPARATORS[comparator]}`

export const options: Options = {
    whereArgumentPath
}
