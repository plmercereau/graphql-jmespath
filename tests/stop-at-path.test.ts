import { it, describe, expect } from 'vitest'
import { JMESPathGraphQL } from '../src/lib'

describe('stop recursing at given paths', () => {
    const expression =
        "countries[].{name:name, statesCount: states[].name | length(@), languages:join(', ', languages[].code)} | sort_by(@, &statesCount) | reverse(@) | [:10]"

    it('should stop recursing a single nested path', () => {
        const result = new JMESPathGraphQL(expression, {
            stopAtPath: 'countries.states'
        }).toString()
        expect(result).matchSnapshot()
    })

    it('should stop recursing an array of paths', () => {
        const result = new JMESPathGraphQL(expression, {
            stopAtPath: ['countries.states', 'countries.languages']
        }).toString()
        expect(result).matchSnapshot()
    })

    it('should stop recursing given a stop function', () => {
        const result = new JMESPathGraphQL(expression, {
            stopAtPath: (path) => path === 'countries.states'
        }).toString()
        expect(result).matchSnapshot()
    })

    it('should return the full query when no stopAtPath option is given', () => {
        const result = new JMESPathGraphQL(expression).toString()
        expect(result).matchSnapshot()
    })

    it('should return the full query when the path to stop at does not exists', () => {
        const result = new JMESPathGraphQL(expression, {
            stopAtPath: 'inexistent'
        }).toString()
        expect(result).matchSnapshot()
    })
})
