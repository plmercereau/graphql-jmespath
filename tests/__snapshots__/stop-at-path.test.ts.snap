// Vitest Snapshot v1

exports[`stop recursing at given paths > should return the full query when no stopAtPath option is given 1`] = `"query { countries { name states { name } languages { code } } }"`;

exports[`stop recursing at given paths > should return the full query when the path to stop at does not exists 1`] = `"query { countries { name states { name } languages { code } } }"`;

exports[`stop recursing at given paths > should stop recursing a single nested path 1`] = `"query { countries { name states languages { code } } }"`;

exports[`stop recursing at given paths > should stop recursing an array of paths 1`] = `"query { countries { name states languages } }"`;

exports[`stop recursing at given paths > should stop recursing given a stop function 1`] = `"query { countries { name states languages { code } } }"`;
